### PR TITLE
fix(client): scope caller-supplied artifact source refs to the owner

### PR DESCRIPTION
**Pull request by julsanchez**
## What & why

The artifact create endpoint (`POST /api/artifacts`) accepted `sessionId`, `sourceQuestId` and `parentArtifactId` from the request body and wrote all three verbatim onto the new artifact, with the caller's own id as the owner and **no authorization check** on the refs. An authenticated user could therefore create an artifact stamped as sourced-from / session-of / child-of another user's session, quest, or artifact that they have no access to - an object-level authorization (IDOR) gap.

The other two callers of the create service (agent runs, notebook import) are unaffected: they pass server-derived refs the caller already holds (the agent run's own session/quest; a freshly created session owned by the importing user), so the guard rejects only the attacker-controlled route.

## The fix

A pure, injectable guard `assertArtifactSourceRefsAccessible(userId, refs, deps)` on the POST route rejects (403 `ForbiddenError`) any supplied ref the caller is not entitled to reference. Each ref is optional and only checked when present.

The bar is **update (write) access, not ownership and not mere read access**, because sessions are shareable and stamping an artifact with a session is a write into that session's graph. A collaborator with **update access** to a shared session legitimately creates artifacts stamped with the owner's `sessionId` (the artifact-persistence paths write the active session's id, which is the owner's on a shared session); an owner-only check would wrongly 403 that flow, and a read-only sharee should not be able to reference the session. So:

- `sessionId` -> allowed iff the caller can **update** the session (owner OR update-share OR group update-share OR global-write share), via `sessionRepository.shareable.findUpdateAccessById(..., { includeGlobalWrite: true })` - the same idiom used for writing into a shared object elsewhere (e.g. `tavern-conversation.ts`). Global-write is an opt-in arm on that method so sharing-mutation callers keep the strict owner/user/group arms.
- `sourceQuestId` -> a quest carries no user of its own, so access is transitive through the session it belongs to (top-level `sessionId`): resolve the quest's session, then apply the same update-access check.
- `parentArtifactId` -> artifacts are **not** shareable (no cross-user artifact flow exists), so this stays a strict owner check: `artifact.userId === caller`.

No dependency on any other in-flight change; uses only methods already on the repos.

## Guide for testers

1. As user A, create an artifact passing a `sessionId` / `sourceQuestId` / `parentArtifactId` that belongs to user B and is **not shared** with A -> expect 403.
2. As user A, create an artifact passing your own session/quest/parent-artifact refs -> expect 201.
3. As user A, create an artifact passing a `sessionId` that user B shared with A **with write/update or global-write access** -> expect 201 (the collaborator flow the guard preserves).
4. As user A, create an artifact passing a `sessionId` that user B shared with A **read-only** -> expect 403 (creating a sourced-from artifact is a write; read-only does not grant it).
5. Create an artifact with none of those refs -> expect 201 (unchanged).

## Regression checks

- Agent-run artifact persistence and notebook-import artifact creation still succeed (they pass caller-held refs).
- Shared-session collaboration: a collaborator with **update** access creating artifacts in a session shared with them is not blocked.
- Creating an artifact with no source refs is unaffected.

## What NOT to test

- Overlay/premium paths - unaffected.
- The read side (listing artifacts by quest) - see the follow-up below; out of scope here.

## Follow-up (separate)

The read side is still owner-unscoped and pre-existing planted refs are not backfilled - tracked in https://github.com/Bike4Mind/bike4mind/issues/2673 (non-blocking for this write-side fix).

## Verify

- `apps/client` typecheck: pass.
- `assertArtifactSourceRefsAccessible.test.ts` (9 cases incl. shared-session-update allowed, stranger blocked, and the quest-session-not-quest-id invariant with `toHaveBeenCalledWith` argument pinning): pass (9/9).

## Checklist

- [x] Types pass
- [x] Tests added and passing
- [x] No new dependencies
- [x] Behavior-only description (no internal identifiers)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by jjmarfa — DISMISSED** [review]
Requesting changes before merge.

**Blocking**

1. `apps/client/pages/api/artifacts/index.ts:104` uses owner-only `sessionRepository.findByIdAndUserId`, but sessions are shareable (`ability.ts:83-107` grants `Permission.update` to sharees). A write-sharee generating an artifact in someone else's session now gets a silent 403 (swallowed by `Promise.allSettled` in `artifactPersistence.ts:188`). Agent runs use a separate server-side flow (`persistRunAsQuest.ts:146`) so this path is live and not superseded. Use a write-capable ownership check instead of owner-only (e.g. `sessionRepository.shareable.findUpdateAccessById`, already the idiom used in `tavern-conversation.ts:244`).
2. The 7 unit tests in `assertArtifactSourceRefsOwned.test.ts` are argument-blind - mutation testing shows all 4 mutants tried survive (7/7 still pass), including swapping the quest's session id for the quest id, which is the exact invariant the util's docblock warns about. Add `toHaveBeenCalledWith` assertions on the injected deps.
3. The two `as` casts in `index.ts:102-105` are provably unneeded (removing them still typechecks) and provably hide a rename (renaming the underlying field with the casts still in place produces no type error, only a runtime 403 for everyone).

**Minor (non-blocking, worth a follow-up commit)**

- Add a route-level POST test; the `@bike4mind/database` mock in `__tests__/index.test.ts:29-33` is missing `sessionRepository`/`questRepository` exports, so those lambdas never actually run today.
- A junk `sessionId`/`sourceQuestId` currently 404s via Mongoose `CastError` rather than a clean 403/422.
- A soft-deleted ref 403s even its own owner (`mongo.ts:457-462` excludes soft-deleted docs from the ownership lookup).

**Follow-up (separate ticket, not blocking this PR)**

The read side is still unscoped: `findByQuestIds` (`ArtifactModel.ts:370-380`) has no owner filter, and `linkNodeArtifacts.ts:52,71` persists the join onto the victim's node, so artifacts planted before this deploy stay attached. No backfill included here.

Core ownership-chain logic in the new guard is correct against the real schemas, and the "other two existing callers unaffected" claim checks out against current main.

**Review by ken-b4m — DISMISSED** [review]
## Round 2 - Request changes (1 P1, 5 P3)

Reviewed at `109e1202`, three commits past the round-1 head. The guard's ownership-chain logic is correct and the round-1 blockers are largely closed: the owner-only lookup is gone, the tests now pin their arguments (all six mutants I tried are killed by a named assertion), and the two casts are verifiably removed. The one blocking item is that the replacement predicate is **narrower than the Session write shape this repo defines elsewhere**, so it 403s a class of legitimate collaborator - and the caller swallows the rejection.

**To unblock:** decide the `isGlobalWrite` question in P1. Everything else is P3 and explicitly **not** worth a push on its own (see the note on push cost at the foot).

---

### P1 - The artifact gate omits the `isGlobalWrite` arm that the session write path grants

`apps/client/pages/api/artifacts/index.ts:107` gates on `sessionRepository.shareable.findUpdateAccessById`, which is three arms - owner, `users[].permissions $in ['update']`, `groups[].permissions $in ['update']` (`packages/database/src/models/content/SharableDocumentModel.ts:131-140`). It has no `isGlobalWrite` arm.

But `isGlobalWrite` **is** part of update access for a Session, independently declared in two places:

- `apps/client/server/auth/ability.ts:88` - `allow(Permission.update, resource, { isGlobalWrite: true })`
- `packages/database/src/utils/ability.ts:77` - the same grant

And the session-append path deliberately honors it, saying so in its own comment (`apps/client/server/managers/sessionCrud.ts:112-124`): "honor the Session write shape (owner + global-write + user/group shares - see ability.ts) exactly as the update path does". It uses `accessibleBy(ability, Permission.update)`, which picks up all four arms.

So for a session with `isGlobalWrite: true`, a collaborator can append messages to it but cannot persist an artifact from it.

**Why this is not merely a 403.** `persistArtifacts` in `apps/client/app/utils/artifactPersistence.ts` wraps the per-artifact posts in `await Promise.allSettled(...)`, so the rejection produces a `console.error` and nothing else. No toast, no retry, no failed state - the artifact is silently gone and the user has no signal to act on. That is what moves this above a permissions nit.

**Aggravating: the arm that was added appears to be unreachable.** The live session share UI hard-codes its grant at `apps/client/app/components/common/ShareModal.tsx:134-136`:

```ts
// Keep this in-case we revisit permission updating on sharing
const [permissions] = useState<{ value: Permission[]; error?: string | null }>({
  value: [Permission.read, Permission.share],
  error: null,
});
```

That is the modal actually mounted (`CombinedNotebooks.tsx:826`, `Files/Browser/Content.tsx:930,943`, `Session/SidenavItem.tsx:305`). The other `ShareModal` - the one with the write toggle - has no importer; `grep "components/ShareModal'"` returns nothing. So no session sharee ever holds `users[].permissions: ['update']`, and the `users[].update` arm this PR now relies on matches nothing in production today. `isGlobalWrite` is settable, via `PUT /api/[type]/[id]/updateSharing` (`apps/client/pages/api/[type]/[id]/updateSharing.ts`, which accepts `isGlobalWrite` for `sessions`) - i.e. the one arm that is reachable is the one the guard omits.

**Severity hinge, which you hold and I do not:** if no production session carries `isGlobalWrite: true`, this is a P2 - a latent divergence to fix before that flag gets used. If any do, it is live silent data loss for their collaborators. Worth a quick count before deciding.

**Where the fix belongs.** Not inline in the route: `canUpdateShareable` in `b4m-core/common/src/utils/shareableAccess.ts` has the identical three-arm gap *and* a docblock stating it mirrors `findUpdateAccessById`. Whichever way this resolves, those two and the route should end up sharing one predicate, or the next reader inherits three subtly different definitions of "can update a session".

<details>
<summary><b>Round-1 blockers - grading</b></summary>

**#1 (owner-only vs shareable sessions) - PARTIALLY CLOSED.** The named mechanism is fixed: `findByIdAndUserId` is gone, replaced by `shareable.findUpdateAccessById`, and that closes the owner-only case for `users[]`/`groups[]` update grants. The consequence you described - "a write-sharee generating an artifact in someone else's session now gets a silent 403 (swallowed by `Promise.allSettled`)" - survives unchanged for `isGlobalWrite` sessions, which is the P1 above. Your diagnosis of the swallowing was right and still applies.

**#2 (argument-blind tests) - FULLY CLOSED.** `assertArtifactSourceRefsAccessible.test.ts` now pins arguments where it matters: `toHaveBeenCalledWith('s1'/'q1'/'a1')` (:32-34), `canUpdateSession` asserted against `'resolved-session'` with `not.toHaveBeenCalledWith('q-raw')` (:61-62) - which is precisely the quest-id-for-session-id mutant you flagged - and `not.toHaveBeenCalled()` for the missing-quest short circuit (:82). I re-ran the mutation set statically: all six mutants I could construct are killed, and for each I identified the specific assertion that kills it rather than assuming the suite went red for the right reason.

**#3 (unneeded casts) - FULLY CLOSED.** Verified against `git show 4a340a8fd:apps/client/pages/api/artifacts/index.ts`: both casts are gone and the closures now read the fields directly.

**Minor - soft-deleted ref 403s its own owner - CLOSED**, as a side effect of the predicate swap. `findUpdateAccessById` has no `deletedAt` filter, so the soft-delete exclusion you traced to `mongo.ts` is no longer on this path.

**Minor - junk ref status code.** You were right and I want to correct the record in case it comes up again: it is a **404**, not a 500. `apps/client/server/middlewares/errorHandler.ts:20-22` maps a Mongoose `CastError` with `path === '_id'` to `NotFoundError`. Still not the clean 403/422 you asked for; carried below as P3.

**Follow-up (read side) - correctly still open.** `findByQuestIds` (`packages/database/src/models/content/ArtifactModel.ts:370`) still has no owner filter, and its one production caller `apps/client/server/questmaster/v5/linkNodeArtifacts.ts:52` does no user scoping. That caller is what makes this PR's premise real, and it is the right thing to have split out. No backfill here, as you said.
</details>

<details>
<summary><b>P3 - worth folding into a later commit, not worth a push on their own</b></summary>

**1. No test exercises the route's real-repo binding.** `apps/client/pages/api/artifacts/__tests__/index.test.ts:24` has `post: () => chain,`, which discards the POST handler entirely, and the `vi.mock('@bike4mind/database')` at :29-33 exports only the three artifact repositories - no `sessionRepository`, no `questRepository`. Only GET describes exist (:52). So the three closures at `index.ts:107` - the actual wiring this PR adds - are executed by nothing. The unit tests cover the guard's logic well; what is untested is that the route hands it the right lookups. This is your round-1 minor, unchanged.

**2. Duplicate test case.** `assertArtifactSourceRefsAccessible.test.ts:37-45` overrides `canUpdateSession: vi.fn(async () => true)` at :40, which is byte-identical to the default at :10, making the case a duplicate of :26-35. Harmless, but it reads as if it intends to vary something.

**3. The PR body no longer matches the code.** It describes `findAccessibleById` and read access, and tester step 3 says a read share should yield 201. The code now requires *update* access, so a read-only sharee gets a 403 - which is the correct behavior and is pinned by `SharableDocumentModel.integration.test.ts:85-89`. A tester following step 3 as written will file a false bug.

**4. `projectId` / `organizationId` are the same class of unchecked caller-supplied ref.** `CreateArtifactSchema` accepts both and `b4m-core/services/src/artifactService/create.ts:141-142` writes them verbatim alongside the three refs this PR now guards. They are latent rather than exploitable: `list.ts:50` forces `userId` into the filter and `get.ts:41`'s `canUserAccessArtifact` has no org/project arm, so nothing reads them for access. I would fold this into the existing read-side follow-up rather than open anything new - the moment a read path does grow an org or project arm, these become the same bug.

**5. Junk `sessionId`/`sourceQuestId` yields 404, not 403/422.** Per the correction above. Cosmetic; the request is still refused.
</details>

<details>
<summary><b>Merge state and CI</b></summary>

CI is green at `109e1202` on every check that has reported: Type Check, Lint, Core Build, `Run Tests (client 1/3)`, `(client 2/3)`, `(client 3/3)`, `client-integration`, `data`, `misc`, `services`, `cli-slack`, and CLI e2e. `Build amd64` was still in flight at my last read, so the verdict does not rest on it.

**An operational note that is easy to lose:** the round-1 `CHANGES_REQUESTED` was submitted against `4a340a8f` and still holds `reviewDecision` at the current head. The branch ruleset has both `dismiss_stale_reviews_on_push` and `require_last_push_approval`, and stale-review dismissal applies to **approvals only** - a `CHANGES_REQUESTED` survives every push. So no amount of pushing clears it; it needs @jjmarfa to re-review or dismiss. That is independent of this review.

**On push cost:** because `require_last_push_approval` is on, every push invalidates the last approval and costs another review cycle. That is why the P3s above are marked as not worth a push individually - land them with the P1 fix, or leave them.

I could not execute the test suite here (`node_modules` is not present in this checkout), so the mutation analysis and every test-vacuity claim above is static reasoning over the source, not a recorded run. Findings are otherwise verified against the working tree at `109e1202`.
</details>

**Comment by julsanchez** [comment]
Thanks for the careful review - all three blockers are addressed in commit 109e12026 (on top of the access-based rename in 5ba408369).

**1. Write-access, not owner-only.** Switched the session check to `sessionRepository.shareable.findUpdateAccessById` (owner or update-share), matching the `tavern-conversation.ts` idiom you pointed to - a write-sharee generating in a shared session now passes instead of getting a 403 swallowed by `Promise.allSettled`. Quest access is transitive through the quest's session and uses the same predicate. `parentArtifactId` stays owner-only (artifacts aren't shareable).

**2. Argument-blind tests.** Added `toHaveBeenCalledWith` assertions pinning each ref against its own id, plus a dedicated test that the session check runs against the quest's **resolved** session id and **never** the raw quest id - the exact mutant you flagged. 9 cases now, and I confirmed the swap fails the suite.

**3. Casts.** Removed both `as` casts. The route now typechecks against the real repo return types, so a field rename fails the build instead of silently 403ing everyone. (This also surfaced that `questRepository.findById` returns the full doc, so `.sessionId` is real at runtime.)

**Minor:** the `CastError`-vs-422 and soft-deleted-owner points are noted. The read-side scoping (`findByQuestIds` + backfill of pre-planted refs) is tracked as a separate follow-up since the backfill is a data migration: https://github.com/Bike4Mind/bike4mind/issues/2673

Re-review when you have a moment - happy to add the route-level POST test too if you'd like it in this PR rather than the follow-up.

**Comment by wescarda** [comment]
## MCP Playwright Test — PR #2662

**URL:** https://app.pr2662.preview.bike4mind.com
**Date:** 2026-09-11
**Method:** Direct API testing with two independent test users (this PR is a pure backend/API authorization guard with no UI surface — verified via `POST /api/artifacts` + session-sharing calls, not browser click-through)

---

## Verdict

> ❌ **FAIL — the update-shared-session collaborator flow is broken.** The IDOR fix itself works (stranger/unshared refs correctly 403, read-only shares correctly 403), but the exact regression case the PR explicitly calls out as preserved — a collaborator with genuine update access to a shared session creating an artifact stamped with that session — is incorrectly rejected with 403.

---

## Summary table

| # | Test step | Result |
|---|---|---|
| 1 | `sessionId` owned by unshared user B → expect 403 | ✅ Pass |
| 1 | `parentArtifactId` owned by unshared user B → expect 403 | ✅ Pass |
| 2 | Own session refs → expect 201 | ✅ Pass |
| 3 | `sessionId` shared by B with **update** access → expect 201 | ❌ FAIL — got 403 `"You do not have access to the referenced session"` |
| 4 | `sessionId` shared by B **read-only** → expect 403 | ✅ Pass |
| 5 | No source refs → expect 201 | ✅ Pass |
| — | Control: total stranger, no invite at all → PUT-rename 404, artifact-create 403 | ✅ Pass (sanity check) |

---

## What works

- Unrelated/unshared `sessionId` and `parentArtifactId` refs are correctly rejected (403) — the core IDOR gap is closed.
- Read-only session shares are correctly blocked from referencing the session (403), confirming update-vs-read is distinguished in that direction.
- Own-refs and no-refs baseline cases are unaffected (201).
- Control test with a total stranger (no relationship to the session at all) behaves consistently, ruling out a broadly-broken auth system.

## Other observations (not blockers)

- `sourceQuestId` scenarios were not independently exercised (no direct quest-creation API; would require a chat/agent run to seed a quest). `sessionId` and `parentArtifactId` were tested directly and follow the same code path.
- Reproduced the Step 3 failure twice, end-to-end, with fresh users both times. Proved user A has genuine update access to the session (a plain `PUT /api/sessions/{id}` rename succeeds as A), yet the new artifact-create guard still 403s the same user/session pair.
- A static-code read-through could not pinpoint the exact line-level cause (the guard's call site and the existing session-update route both appear to call `sessionRepository.shareable.findUpdateAccessById` identically) — this looks like a live wiring/runtime issue, not something visible from source alone. Recommend server-side logging of `req.user.id` and the exact `sessionId` at the guard call site to pin it down. Worth noting the PR's own unit tests mock `canUpdateSession` entirely, so this real wiring was never integration-tested before merge.

No screenshots — this PR has no UI surface (pure backend/API authorization change).

**Comment by julsanchez** [comment]
Pushed 13e0c883c, which closes @ken-b4m's round-2 **P1** (the missing `isGlobalWrite` arm).

@jjmarfa - your round-1 blockers (owner-only lookup, argument-blind tests, the two casts) were addressed earlier in 109e12026 and answered in the comment above; nothing here re-opens them. This push only closes the `isGlobalWrite` remnant of your blocker 1 that @ken-b4m elevated to P1. A re-review to clear your standing `CHANGES_REQUESTED` would help - it survives pushes and blocks merge on its own.

Below is what changed for ken's P1 and the P3 decisions.

## P1 (missing `isGlobalWrite` arm) - fixed

The artifact source-ref guard now accepts the global-write arm. Implementation: an opt-in `includeGlobalWrite` flag on `findUpdateAccessById` (`SharableDocumentModel.ts`), which the artifact-create `canUpdateSession` closure passes (`artifacts/index.ts`). Resulting matrix: owner / `users[].update` / `groups[].update` / **`isGlobalWrite`** -> 201; read-only + stranger -> 403.

**Why a flag, not the arm unconditionally.** `findUpdateAccessById` is shared by the sharing-mutation callers (`sharingService/updateDocumentSharing.ts`, `pages/api/[type]/[id]/updateSharing.ts`), and CASL grants `isGlobalWrite` only `Permission.update`, never `share`/`delete` (`ability.ts:88` plus the per-permission loop at `:90`). Adding the arm unconditionally would let a global-write sharee re-share or revoke - an escalation. The flag keeps every existing caller on the strict three arms (default `false`) and adds the arm only where a session write actually happens.

**On "three definitions should share one predicate."** With the flag defaulting off, `findUpdateAccessById`'s default is still exactly the three arms, so it and `canUpdateShareable` (`shareableAccess.ts`) still mirror each other - the docblock invariant holds. The opt-in is one explicit deviation at a single call site, not a third silent definition. If you would rather fully unify (a shared `canWriteSession` that the route and `canUpdateShareable` both route through, with global-write folded in), I am happy to - I kept it minimal to avoid disturbing the sharing-mutation semantics.

**Severity hinge (prod count of `isGlobalWrite: true` sessions).** I cannot run that query from here. The fix is correct either way - it removes the latent divergence whether or not any such sessions exist today - but a prod count would settle P2-vs-live-data-loss if you want it before merge.

## Regression coverage

New real-Mongo integration test `SessionModel.findUpdateAccessById.integration.test.ts`: asserts owner/user-update/group-update match; global-write excluded by default and included only on opt-in; read-only sharee + stranger rejected even with the flag on. This exercises the real `findUpdateAccessById` arms against Mongo - the layer the bug lived in, which the mocked guard unit test could never catch (ken P3.1 / jjmarfa minor).

It does not stand up a full route POST: `post: () => chain` plus the partial `@bike4mind/database` mock in `__tests__/index.test.ts` still discard the handler, so the route wiring is the one-line closure change plus the guard unit test. If you want the full route POST test too I will add it - flagging it is the heavier lift you both marked "not worth a push on its own."

## P3s not taken (bundled, per the push-cost note)

- Duplicate guard test case (`:37-45` vs the `:10` default) - pre-existing, harmless; left as-is.
- `projectId` / `organizationId` unchecked refs - agreed; folding into the read-side follow-up (latent until a read path grows an org/project arm).
- Junk ref -> 404 rather than 403/422 - cosmetic; request still refused; left for the follow-up.
- PR-body copy (P3.3) - the read-access wording you flagged was already corrected in an earlier commit (the body describes *update* access, and step 3 already expects 201 for write/update and 403 for read-only). I additionally added the global-write arm to the `sessionId` description and step 3 so the body matches the new behavior.

## Review state

Both `CHANGES_REQUESTED` survive this push (the ruleset dismisses stale approvals only), so clearing them needs a re-review from each of you - jjmarfa's blockers are already closed, ken's P1 is the item addressed here. QA re-test delta posted separately for @wescarda.

**Comment by julsanchez** [comment]
@wescarda - re-test delta since your 2026-09-11 04:10 pass (which ran against 109e12026; the fixed head is now 13e0c883c).

**Preview is stale.** Your run hit 4a340a8; it must be redeployed to 13e0c883c before re-testing.

## Root cause of your Step 3 FAIL - exactly what you suspected

The guard's session check delegated to `findUpdateAccessById`, whose arms are owner + `users[].update` + `groups[].update`, with **no `isGlobalWrite` arm**. The share UI grants write only via the global-write flag, so your User A (genuine update access, `PUT /api/sessions/{id}` rename succeeds) held `isGlobalWrite` - the one arm the guard dropped -> 403. Your note that "rename succeeds but artifact-create 403s the same pair, and both call the same method" pointed straight at it: same method, but it omitted the arm A actually held. Your point that the PR's unit tests mocked `canUpdateSession` entirely (so the real wiring was never integration-tested) is now addressed - see the new Mongo test below.

## What changed (13e0c883c)

The artifact guard now opts into the global-write arm.

| Your step | Before | Now expected |
|---|---|---|
| **3** - `sessionId` shared by B with update / global-write access, A creates artifact | 403 FAIL | **201** |
| 1 - unshared B's `sessionId` / `parentArtifactId`, A | 403 | 403 (unchanged) |
| 4 - read-only share, A | 403 | 403 (unchanged) |
| 2 - own refs | 201 | 201 (unchanged) |
| 5 - no refs | 201 | 201 (unchanged) |
| control - total stranger | rename 404 / create 403 | unchanged |

## How to re-test

- **Must re-run (the behavior that changed): Step 3.** A with a genuine write share to B's session creates an artifact stamped with that `sessionId` -> now expect **201**. Please record HOW A's access is granted (global-write via `updateSharing`, vs any directly-seeded `users[].update`), since that identifies the exact arm exercised - global-write is the one this fix adds.
- **Please re-confirm (highest-value regression guard): Step 4.** A read-only share must STILL 403. The fix adds global-write only, never read - this is the arm we most want proof we did not loosen.
- **Can skip (unchanged, code untouched): Steps 1, 2, 5, control.** Same code path, already green, unaffected. Re-run only for a full sweep.
- **Still not independently exercised: `sourceQuestId`.** As you noted, no direct quest-creation API. It routes through the same `canUpdateSession` predicate as `sessionId` (via the quest's session), so a green Step 3 covers it in principle; a chat/agent-seeded quest would confirm end-to-end if you can seed one, but it is not a new gap from this change.

**Review by jjmarfa — COMMENTED** [review]
## Review summary

This closes out the session-sharee silent-403 IDOR finding from earlier rounds: the route now calls `sessionRepository.shareable.findUpdateAccessById` (apps/client/pages/api/artifacts/index.ts:111), the argument-blind unit tests now catch a swapped-argument mutant, and the unneeded `as` casts were removed. Verified directly: guard unit tests pass, new integration suite passes, route tests pass, `apps/client` typecheck is clean, `pnpm lint:check` is clean, and CI is green on this SHA.

No blocking issues. A few nits, none of which need to hold up merge:

1. **`includeGlobalWrite` opt-in arm** (`b4m-core/db-core` `SharableDocumentModel.ts:131-149`, widened signature in `@bike4mind/common`'s `ShareableDocumentTypes.ts:118`). Traced every live caller that could exercise it and it looks inert: session reads have no global-write arm (`sessionService/get.ts:24`), the chat path is stricter (`ChatCompletionInvoke.ts:147`), and a `users[]`-update sharee already passes without it. Consider dropping it (and the two shared-package edits) unless there's a real flow it's meant to unblock -- would shrink the PR back to `apps/client` only.
2. `shareableAccess.ts:8`'s "mirrors findUpdateAccessById's three arms" comment is now stale given the new arm -- worth a one-line update regardless of (1).
3. The route test's `@bike4mind/database` mock (`apps/client/pages/api/artifacts/__tests__/index.test.ts`) still doesn't export `sessionRepository`/`questRepository`, and POST is stubbed as a no-op passthrough, so there's no route-level 403 test. Not blocking, but a swapped field-mapping mutant on the route wiring survives -- worth a follow-up test.
4. Minor, non-blocking edge cases: a soft-deleted session ref silently 403s its own owner (via the `pre('findOne')` filter); a junk ref id 404s via CastError rather than 403/422; `projectId`/`organizationId` in the same request schema remain unguarded (no live read path found, so latent only).

Out of scope for this PR, tracked separately (issue #2673): the read side (`ArtifactModel.findByQuestIds`) is still unscoped by owner, so artifact refs planted before this fix stay attached on read. No backfill needed here.

Nice work closing the sharee gap cleanly with a real repo idiom instead of a bespoke check.

File: apps/client/pages/api/artifacts/index.ts [hunk 1]
```diff
@@ -84,6 +91,29 @@ const handler = baseApi()
...
+          parentArtifactId: validatedData.parentArtifactId,
+        },
+        {
+          // Include the global-write share arm: a global-write sharee may write into the session
+          // graph (stamp an artifact with its id), matching the CASL update ability. Owner and
+          // update/group-update shares still pass; read-only sharees and strangers still 403.
+          canUpdateSession: async id =>
+            !!(await sessionRepository.shareable.findUpdateAccessById(req.user!, id, { includeGlobalWrite: true })),
+          getQuestSessionId: async id => (await questRepository.findById(id))?.sessionId ?? null,
+          getArtifactOwner: async id => (await artifactRepository.findOne({ id }))?.userId ?? null,
```
Comment (vinchi777) [inline apps/client/pages/api/artifacts/index.ts:113]: **P2** - the three closures here are the only untested part of the fix.

`assertArtifactSourceRefsAccessible.test.ts` mocks all three deps away (correctly - it is testing the policy), and the integration test covers `findUpdateAccessById`'s arm-set. What nothing covers is this glue, and the route's own test file `apps/client/pages/api/artifacts/__tests__/index.test.ts` only captures the GET handler (`post: () => chain`), so POST is never exercised at all.

That matters most for `getArtifactOwner`: `artifactRepository.findOne({ id })` is right because `id` is a real String path on `ArtifactSchema` - but `ArtifactRepository` inherits `findById` from `BaseRepository`, which looks up `_id`. A future edit swapping one for the other type-checks, passes every test in this PR, and silently 403s every legitimate `parentArtifactId`. Same for `questRepository.findById(id)?.sessionId` if the quest lookup ever moves to the custom-id idiom.

Not blocking - every branch fails closed, so a wiring slip is a denial, not a bypass. But a single route-level test that drives POST with stub repositories and asserts 403-vs-201 for each ref would pin the part most likely to drift.

File: apps/client/server/utils/assertArtifactSourceRefsAccessible.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+ * owner's `sessionId` (the artifact-persistence paths write the active session's id); an owner-only
+ * check would wrongly 403 that flow, and a read-only sharee should not be able to reference the
+ * session either. `canUpdateSession` therefore mirrors the session repo's update-access predicate
+ * (owner OR update share OR group update share OR global-write share) - the same idiom used
+ * elsewhere for writing into a shared object - and closes over the caller in the route.
+ *
+ * A quest carries no user of its own, so its access is transitive through the session it belongs
+ * to (top-level `sessionId`): resolve the quest's session, then apply the same update-access check.
+ *
+ * Artifacts are NOT shareable (no cross-user artifact flow exists), so `parentArtifactId` stays a
```
Comment (vinchi777) [inline apps/client/server/utils/assertArtifactSourceRefsAccessible.ts:25]: **P3** - the owner-only check for `parentArtifactId` is the right call, but the stated reason ("no cross-user artifact flow exists") is not accurate and will mislead whoever revisits this.

Artifacts are not `ShareableDocument`s, true - but they carry their own `permissions.canRead/canWrite/canDelete` arrays and a `visibility` enum, and both `ArtifactRepository.findByUserWithAccess` and `canUserAccessArtifact` in `artifactService/get.ts` grant non-owners read *and* write on that basis. So a cross-user artifact access flow does exist; it just is not the shareable one.

Suggest narrowing the claim to what is actually load-bearing - something like "artifacts are not `ShareableDocument`s, so there is no session-style update-share to honour here; parenting stays owner-only until a concrete cross-user parenting flow needs it."

**Review by vinchi777 — APPROVED** [review]
## PR Review: fix(client): scope caller-supplied artifact source refs to the owner

### **APPROVE** — the IDOR is closed correctly and the access bar is the right one (reviewed at `13e0c883`)

| Severity | Finding | Location |
|----------|---------|----------|
| P2 | The guard's three adapter closures have no test coverage — the route's own test file captures only the GET handler | `apps/client/pages/api/artifacts/index.ts:110` |
| P3 | Doc comment's justification for the owner-only parent check is inaccurate: a cross-user artifact access flow does exist | `apps/client/server/utils/assertArtifactSourceRefsAccessible.ts:25` |

The shape of this is right. I verified the three claims the fix rests on:

- **`artifactService.create` has exactly three callers** (this route, `persistAgentArtifacts.ts`, `notebookImportComplete.ts`), and only this one takes the refs from a request body — so putting the guard on the route rather than in the service does not leave a hole today.
- **There is no second write path.** `artifactService.update`'s schema does not accept `sessionId` / `sourceQuestId` / `parentArtifactId`, so an attacker cannot create a ref-free artifact and then PATCH the refs in.
- **`includeGlobalWrite` is not a widening of the shared primitive.** `isGlobalWrite` lives on the base shareable schema (`SharableDocumentModel.ts:46`, default `false`), so every shareable model carries the path and the new `$or` arm can never degrade to an empty `{}` match; off by default, sharing-mutation callers keep the strict arms. The integration test pins exactly that split against real Mongo, which is the right place to pin it — the pure-guard unit test mocks `canUpdateSession` away.

Update-access rather than ownership is the correct bar, and the reasoning in the PR body holds up: `useStreamingArtifactPersistence` stamps `quest.sessionId`, which on a shared session is the owner's id, so an owner-only check would have 403'd the collaborator flow. Every branch of the guard fails closed (a missing quest or artifact rejects rather than falling through).

Scope is well managed: the read side (`ArtifactRepository.findByQuestIds`, which has no owner filter) is genuinely the thing that gives a planted ref its impact, and it is correctly split out into #2673 — I confirmed that issue is open and describes both the read scoping and the backfill.

### P2 — Important (non-blocking)

1. **No route-level coverage for the adapter wiring.** See the inline comment on `index.ts:110`.

### P3 — Minor

1. **Inaccurate justification in the doc comment.** See the inline comment on `assertArtifactSourceRefsAccessible.ts:25`.

### What's done well

- The injectable-deps shape keeps the policy decision pure and unit-testable while the I/O stays at the route; the nine unit cases pin argument identity (`toHaveBeenCalledWith('resolved-session')` plus `not.toHaveBeenCalledWith('q-raw')`), which is exactly the regression that would otherwise escape.
- Layering the real-Mongo integration test under the mocked unit test is the right call — the arm-set is the part a mock cannot vouch for, and the test explicitly covers the read-only sharee and the stranger with the flag on.
- The behavior change for a read-only sharee (now 403) is a deliberate, documented tightening rather than an accident, and it is called out in the tester guide.

**Comment by wescarda** [comment]
## MCP Playwright Test — PR #2662 (re-test: Steps 3 & 4 vs fixed head)

**URL:** https://app.pr2662.preview.bike4mind.com
**Date:** 2026-09-11
**Method:** Direct API testing with two independent test users (this PR is a pure backend/API authorization change, no UI surface). Preview was stale (`4a340a8`) at the start of this run; redeployed to `13e0c883c` before testing.

---

## Verdict

> ✅ **PASS — the `isGlobalWrite` fix in `13e0c883c` works as intended.** Step 3 (global-write share) now returns `201` as expected; Step 4 (global-read-only share, same session/user) still correctly returns `403`. Root cause matches your diagnosis exactly.

---

## Summary table

| # | Test step | Result |
|---|---|---|
| 1 | Preview redeployed to `13e0c883c` | ✅ Pass — confirmed before testing |
| 2 | Step 3 — session shared `isGlobalWrite:true`, non-owner A creates artifact with that `sessionId` | ✅ Pass — `201` |
| 3 | Step 4 — same session flipped to `isGlobalRead:true`/`isGlobalWrite:false`, same A retries | ✅ Pass — `403 "You do not have access to the referenced session"` |

---

## What works

- The new opt-in `includeGlobalWrite` arm on `findUpdateAccessById`, wired into the artifact-create guard's `canUpdateSession`, correctly grants a global-write sharee the ability to reference the session's id on a new artifact.
- Flipping the same session to global-read-only correctly reverts to 403 for the same caller — confirms the fix adds write access without loosening the read-only boundary (the regression guard you flagged as highest-value).

## Other observations (not blockers)

- Per your guidance, Steps 1, 2, 5, and the stranger control were not re-run (unchanged code path, already green previously). `sourceQuestId` still not independently exercised (no direct quest-creation API; shares the same `canUpdateSession` predicate as `sessionId`).

---

No screenshots (no UI surface on this PR).
